crossopt/assignment4/countislands.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+
+
+# disjoint set union for grid components
+class _GridDSU:
+    def __init__(self, n, m):
+        self.parent = dict()
+        self.size = dict()
+        for i in range(n):
+            for j in range(m):
+                self.parent[(i, j)] = (i, j)
+                self.size[(i, j)] = 1
+
+    def __getitem__(self, u):
+        if self.parent[u] != u:
+            self.parent[u] = self[self.parent[u]]
+        return self.parent[u]
+
+    def join(self, u, v):
+        u = self[u]
+        v = self[v]
+        if self.size[u] < self.size[v]:
+            u, v = v, u
+        self.size[u] += self.size[v]
```

Comment:
Ok, my code for this assignment was really awkward to look at several days later. And I still think that when I see it next time I'll be able to simplify it again:)

I don't need to set size[v], because in my implementation of the DSU, I only ever need to know the correct size of a vertice that is the root of its tree, and v is joined to u and as such is not a root